### ISSUE TITLE
MaxDepth Query Analysis Raising Variable Error

### ISSUE THREAD

**Issue by sj26** [rmosolgo/graphql-ruby#196]
We found a weird issue recently while testing some queries. When executing a query with missing variable values which causes a validation error coupled with a max depth analyzer we see the analyzer try to access the `GraphQL::Query#variables` anyway which raises a `GraphQL::Query::VariableValidationError` external to the execution's error handling.

I've created a breaking case on a branch: https://github.com/sj26/graphql-ruby/compare/analyzer-directive-variable-error

Here's what happens on run:

```
GraphQL::Analysis::MaxQueryDepth::with a query including a directive and variables without values
  test_0001_is broken                                            ERROR (0.00s)
GraphQL::Query::VariableValidationError:         GraphQL::Query::VariableValidationError: Variable id of type Int! was provided invalid value
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query/variables.rb:32:in `get_graphql_value'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query/variables.rb:10:in `block in initialize'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query/variables.rb:8:in `each'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query/variables.rb:8:in `each_with_object'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query/variables.rb:8:in `initialize'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query.rb:85:in `new'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query.rb:85:in `variables'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query.rb:113:in `arguments_for'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query/directive_resolution.rb:7:in `block in include_node?'
            /usr/local/rbenv/versions/2.3.1/lib/ruby/2.3.0/set.rb:306:in `each_key'
            /usr/local/rbenv/versions/2.3.1/lib/ruby/2.3.0/set.rb:306:in `each'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query/directive_resolution.rb:5:in `include_node?'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/query_depth.rb:32:in `call'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:46:in `block in visit_analyzers'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:43:in `each'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:43:in `visit_analyzers'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:33:in `reduce_node'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:36:in `block in reduce_node'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:35:in `each'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:35:in `reduce_node'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:36:in `block in reduce_node'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:35:in `each'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:35:in `reduce_node'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:21:in `block in analyze_query'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:20:in `each'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/analysis/analyze_query.rb:20:in `analyze_query'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query.rb:146:in `analysis_errors'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/query.rb:62:in `result'
            /Users/sj26/Projects/graphql-ruby/lib/graphql/schema.rb:70:in `execute'
            spec/graphql/analysis/max_query_depth_spec.rb:12:in `block (2 levels) in <main>'
```

Sadly I'm not GraphQL-savvy enough yet to supply a fix. 😥

**Comment by rmosolgo** [comment]
Thanks for that reproducing case! I guess this should return something like a validation error...

**Comment by rmosolgo** [comment]
Should be fixed in e0b9cfe can you try master?

**Comment by gjtorikian** [comment]
I think the same issue might be occurring with Relay `id` fields:

```
RuntimeError: Invalid complexity: nil on id\n\t/vendor/gems/2.1.7/ruby/2.1.0/gems/graphql-0.17.2/lib/graphql/analysis/query_complexity.rb:69
```

The GraphQL snippet it's upset about is standard:

``` graphql
              edges {
                node {
                  id
                }
              }
```

I do not have complexities for any of my fields (I assume they all get a default of 1?).

**Comment by rmosolgo** [comment]
Derp, looks like `GlobalIdField` extends `Field` but doesn't call `super` in `#initialize` :S

**Comment by sj26** [comment]
Wonderful, works like a treat, thanks! 👌

**Comment by rmosolgo** [comment]
@gjtorikian Fixed global id fields in 77828ae6fc21c20261e87fbec623260aadaff3c7 (by removing them :rocket: and just using a normal `GraphQL::Field`)

**Comment by gjtorikian** [comment]
Legit! 👍
